File: ahi-bodyscan-react/downloaded/ahi-sdk-bodyscan-android-24.10-dev/PartSegmentation/src/main/cpp/Segmentation.cpp
//
//  AHI
//
//  Copyright (c) AHI. All rights reserved.
//

#include "Segmentation.hpp"

#include "ahiCommon.hpp"
#include "ahiFactoryInspection.hpp"
#include "ahiFactorySegment.hpp"
#include "Common.hpp"

cv::Mat
Segmentation::segment(const cv::Mat &capture, cv::Mat contourMask, BodyScanCommon::Profile profile,
                      std::map<std::string, cv::Point2f> poseJoints, const char *modelBuffer,
                      std::size_t modelBufferSize) {
    std::string profileString = profile == BodyScanCommon::Profile::front ? "front" : "side";
    ahiPoseInfo poseInfoPredictions;
    poseInfoPredictions.CentroidHeadTop = poseJoints.at("CentroidHeadTop");
    poseInfoPredictions.CentroidNeck = poseJoints.at("CentroidNeck");
    poseInfoPredictions.CentroidRightAnkle = poseJoints.at("CentroidRightAnkle");
    poseInfoPredictions.CentroidLeftAnkle = poseJoints.at("CentroidLeftAnkle");
    poseInfoPredictions.CentroidRightKnee = poseJoints.at("CentroidRightKnee");
    poseInfoPredictions.CentroidLeftKnee = poseJoints.at("CentroidLeftKnee");
    poseInfoPredictions.CentroidRightHip = poseJoints.at("CentroidRightHip");
    poseInfoPredictions.CentroidLeftHip = poseJoints.at("CentroidLeftHip");
    poseInfoPredictions.CentroidRightHand = poseJoints.at("CentroidRightHand");
    poseInfoPredictions.CentroidLeftHand = poseJoints.at("CentroidLeftHand");
    poseInfoPredictions.CentroidRightElbow = poseJoints.at("CentroidRightElbow");
    poseInfoPredictions.CentroidLeftElbow = poseJoints.at("CentroidLeftElbow");
    poseInfoPredictions.CentroidRightShoulder = poseJoints.at("CentroidRightShoulder");
    poseInfoPredictions.CentroidLeftShoulder = poseJoints.at("CentroidLeftShoulder");
    ahiSegmentInfo segInfo = ahiSegmentInfo();
    ahiCommon common = ahiCommon();
    // load the tflite model first
    common.loadTensorFlowModelFromBuffer(modelBuffer, modelBufferSize,
                                         "segnet.tflite"); // could also use "segmentnet.tflite"

    // load the image
    common.feedInputBufferImageToCpp(capture.data, capture, "segnet.tflite");
    common.segment(capture, contourMask, poseInfoPredictions, profileString, segInfo);
    return segInfo.segmentMask;
}


std::vector<cv::Mat> Segmentation::segmentAll(
        const std::vector<cv::Mat> &captures,
        std::vector<cv::Mat> contourMasks,
        std::vector<BodyScanCommon::Profile> profiles,
        std::vector<std::map<std::string, cv::Point2f>> poseJoints,
        const char *modelBuffer,
        std::size_t modelBufferSize
) {
    ahiCommon common = ahiCommon();
    // load the tflite model first
    common.loadTensorFlowModelFromBuffer(modelBuffer, modelBufferSize, "segnet.tflite"); // could also use "segmentnet.tflite"
    std::vector<cv::Mat> silhouettes;
    for (int index = 0; index < captures.size(); ++index) {
        std::string profile = profiles[index] == BodyScanCommon::Profile::front ? "front" : "side";
        auto joints = poseJoints[index];
        auto contourMask = contourMasks[index];
        auto capture = captures[index];
        ahiPoseInfo poseInfoPredictions;
        poseInfoPredictions.CentroidHeadTop = joints.at("CentroidHeadTop");
        poseInfoPredictions.CentroidNeck = joints.at("CentroidNeck");
        poseInfoPredictions.CentroidRightAnkle = joints.at("CentroidRightAnkle");
        poseInfoPredictions.CentroidLeftAnkle = joints.at("CentroidLeftAnkle");
        poseInfoPredictions.CentroidRightKnee = joints.at("CentroidRightKnee");
        poseInfoPredictions.CentroidLeftKnee = joints.at("CentroidLeftKnee");
        poseInfoPredictions.CentroidRightHip = joints.at("CentroidRightHip");
        poseInfoPredictions.CentroidLeftHip = joints.at("CentroidLeftHip");
        poseInfoPredictions.CentroidRightHand = joints.at("CentroidRightHand");
        poseInfoPredictions.CentroidLeftHand = joints.at("CentroidLeftHand");
        poseInfoPredictions.CentroidRightElbow = joints.at("CentroidRightElbow");
        poseInfoPredictions.CentroidLeftElbow = joints.at("CentroidLeftElbow");
        poseInfoPredictions.CentroidRightShoulder = joints.at("CentroidRightShoulder");
        poseInfoPredictions.CentroidLeftShoulder = joints.at("CentroidLeftShoulder");
        ahiSegmentInfo segInfo = ahiSegmentInfo();
        // load the image
        common.feedInputBufferImageToCpp(capture.data, capture, "segnet.tflite");
        common.segment(capture, contourMask, poseInfoPredictions, profile, segInfo);
        silhouettes.push_back(segInfo.segmentMask);
    }
    return silhouettes;
}

File: ahi-bodyscan-react/downloaded/ahi-sdk-bodyscan-android-24.10-dev/PartSegmentation/src/main/java/com/advancedhumanimaging/sdk/bodyscan/partsegmentation/Segmentation.kt
//
//  AHI
//
//  Copyright (c) AHI. All rights reserved.
//

package com.advancedhumanimaging.sdk.bodyscan.partsegmentation

import android.content.Context
import android.graphics.Bitmap
import android.graphics.PointF
import com.advancedhumanimaging.sdk.bodyscan.common.BodyScanError
import com.advancedhumanimaging.sdk.bodyscan.common.Profile
import com.advancedhumanimaging.sdk.bodyscan.common.interfaces.AHIBSResourceType
import com.advancedhumanimaging.sdk.bodyscan.common.interfaces.IResources
import com.advancedhumanimaging.sdk.bodyscan.common.interfaces.ISegmentation
import com.advancedhumanimaging.sdk.bodyscan.partsegmentation.jni.SegmentationJNI
import com.advancedhumanimaging.sdk.common.models.AHIResult
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object Segmentation : ISegmentation {
    private val expectedJoints = listOf(
        "CentroidHeadTop",
        "CentroidNeck",
        "CentroidRightAnkle",
        "CentroidLeftAnkle",
        "CentroidRightKnee",
        "CentroidLeftKnee",
        "CentroidRightHip",
        "CentroidLeftHip",
        "CentroidRightHand",
        "CentroidLeftHand",
        "CentroidRightElbow",
        "CentroidLeftElbow",
        "CentroidRightShoulder",
        "CentroidLeftShoulder",
        "CentroidNose"
    )

    override suspend fun segment(
        capture: Bitmap,
        contourMask: Bitmap,
        profile: Profile,
        poseJoints: Map<String, PointF>,
        context: Context,
        resources: IResources
    ): AHIResult<Bitmap> {
        return withContext(Dispatchers.IO) {
            if (capture.width != 720 || capture.height != 1280) {
                return@withContext AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENTATION_INCORRECT_CAPTURE_RESOLUTION)
            }
            if (contourMask.width != 720 || contourMask.height != 1280) {
                return@withContext AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENTATION_INCORRECT_CONTOUR_RESOLUTION)
            }
            if (expectedJoints.any { !poseJoints.containsKey(it) }) {
                return@withContext AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENTATION_MISSING_JOINTS)
            }
            val modelBuffer = resources.getResource("segnet", AHIBSResourceType.AHIBSResourceTypeML, context).getOrNull()
            if (modelBuffer != null) {
                val img = SegmentationJNI.segment(
                    capture,
                    contourMask,
                    profile,
                    poseJoints,
                    modelBuffer,
                    modelBuffer.size
                )
                if (img != null) {
                    AHIResult.success(img)
                } else {
                    AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENTATION_FAILED)
                }
            } else {
                AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENTATION_MODEL_MISSING)
            }
        }
    }

    suspend fun segment(
        captures: Array<Bitmap>,
        contourMasks: Array<Bitmap>,
        profiles: Array<Profile>,
        poseJoints: Array<Map<String, PointF>>,
        context: Context,
        resources: IResources
    ): AHIResult<Array<Bitmap>> {
        return withContext(Dispatchers.IO) {
            captures.forEach {
                if (it.width != 720 || it.height != 1280) {
                    return@withContext AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENT_LIST_INCORRECT_CAPTURE_RESOLUTION)
                }
            }
            contourMasks.forEach {
                if (it.width != 720 || it.height != 1280) {
                    return@withContext AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENT_LIST_INCORRECT_CONTOUR_RESOLUTION)
                }
            }
            poseJoints.forEach { joints ->
                if (expectedJoints.any { !joints.containsKey(it) }) {
                    return@withContext AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENT_LIST_MISSING_JOINTS)
                }
            }
            val modelBuffer = resources.getResource("segnet", AHIBSResourceType.AHIBSResourceTypeML, context).getOrNull()
            if (modelBuffer != null) {
                val img = SegmentationJNI.segmentAll(
                    captures,
                    contourMasks,
                    profiles,
                    poseJoints,
                    modelBuffer,
                    modelBuffer.size
                )
                if (img != null) {
                    AHIResult.success(img)
                } else {
                    AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENT_LIST_FAILED)
                }
            } else {
                AHIResult.failure(BodyScanError.BODY_SCAN_SEGMENT_LIST_MODEL_MISSING)
            }
        }
    }
}